feat: add [dojo::interface] attribute

remybar opened the pull request:
Related to DOJ-132, https://github.com/dojoengine/dojo/issues/1458.

Add a new `[dojo::interface]` attribute which is expanded in `[starknet::interface]`, allowing the user to ignore `TContractState` type and `self` parameter.

remybar commented:
Here the first part of the issue related to `[dojo::interface]`.
- As noted in the code, I ignored some function parts (like `nopanic` clause) as, afaik, there are not possible in a `starknet::interface` trait.
- I was wondering if I can improve the indentation of the generated code or if it's not important 🤔

glihm commented on crates/dojo-lang/src/interface.rs:
```diff
@@ -0,0 +1,149 @@
+                .type_clause(db).ty(db).as_syntax_node()
+                .get_text(db).trim().to_string();
+
+            need_to_add_self = !(
+                first_param_name.eq(&"self".to_string()) 
+                && first_param_type.eq(&"@TContractState".to_string())
+            );
```
Wondering if here we should emit a warning (or error?) if the function is using `ref self`.
In dojo, systems are supposed to be 100% stateless. For this reason, we should consider warning/send error in this particular case.

remybar replied:
Yes, good idea !

glihm commented on crates/dojo-lang/src/interface.rs:
```diff
@@ -0,0 +1,149 @@
+    ) -> Vec<RewriteNode> {
+        let mut rewrite_nodes = vec![];
+
+        // Some notes:
+        // * as rewritten functions belong to a starknet::interface:
+        //     - there is no generic parameter,
+        //     - there is no nopanic and implicits clauses.
+        // * there is no function body in a trait.
```
Could we move this in a function documentation instead of notes inside the body?

glihm reviewed (commented):
First quick review of the draft, very nice start!
Left only minor comments for now, good job!

remybar commented:
> First quick review of the draft, very nice start! Left only minor comments for now, good job!

Thanks ser ! 🙏 

I don't know why but if a trait contains other elements than a function, when I copy these elements with `vec![RewriteNode::Copied(el.as_syntax_node())]`, they are wrongly copied 🤔 

For example, I tried with a constant `const ONE: u8 = 1;` and it became `const ONE` ...

https://github.com/dojoengine/dojo/blob/a3dca4ae5dd2fa80bfcdef97c13e9e69753f19fc/crates/dojo-lang/src/interface.rs#L28-L44

tarrencev commented on crates/dojo-lang/src/interface.rs:
```diff
@@ -0,0 +1,197 @@
+                        return system.rewrite_function(db, fn_ast.clone());
+                    }
+                    
+                    //TODO: if the trait body contains other things than functions, 
```
Hm is anything other than functions valid in a trait? If so, we should at least output a diagnostic here to say it is not supported yet

remybar replied:
Ok, I did some tests and if I add a const or an enum in a `[starknet::interface]` trait I got an error. So, let's say we cannot have anything other than functions in a `[starknet::interface]`. 
Yes @tarrencev, I will output a diagnostic in this case 👍

tarrencev commented on crates/dojo-lang/src/interface.rs:
```diff
@@ -0,0 +1,197 @@
+
+pub struct DojoInterface {
+    diagnostics: Vec<PluginDiagnostic>,
+    dependencies: HashMap<smol_str::SmolStr, Dependency>,
```
```suggestion
```

I dont think we need this here

remybar replied:
thanks !

tarrencev commented on crates/dojo-lang/src/interface.rs:
```diff
@@ -0,0 +1,197 @@
+                content: builder.code,
+                aux_data: Some(DynGeneratedFileAuxData::new(DojoAuxData {
+                    models: vec![],
+                    systems: vec![SystemAuxData {
```
Hmm is this necessary?

remybar commented:
I rewrote the code for `dojo:interface` to make it easier to read and to not have to handle corner cases. The idea is to patch the smallest parts of code.

I updated the `contract.rs` file to handle `self` and `world` parameters in free functions and functions inside Impl block.

I still need to look at a weird issue when I add a constant in a trait with the `[starknet::interface]]` attribute. I got a `= 1` at the beginning of the expanded trait code 🤔

Now, I will also have a look at CI errors ;-)

remybar commented:
> I still need to look at a weird issue when I add a constant in a trait with the `[starknet::interface]]` attribute. I got a `= 1` at the beginning of the expanded trait code 🤔 In fact, I'm not sure we can add a constant in a trait (I was not able to compile a Cairo project with this kind of thing), but the parser accepts other things than functions in a trait ... And to have a 100% code coverage for the CI, I need to test that branch with the diagnostic logging 😅
> 

After some tests, I found that if I add `const ONE: u8 = 1;` at the beginning of a trait, `=1;` is parsed as a [trait function keyword](https://docs.rs/cairo-lang-syntax/latest/cairo_lang_syntax/node/ast/struct.FunctionDeclaration.html).
So, I searched directly in the cairo parser, and it appears that the expected pattern is `const <name>: types;` (see [here](https://github.com/starkware-libs/cairo/blob/600553d34a1af93ff7fb038ac677efdd3fd07010/crates/cairo-lang-parser/src/parser.rs#L835)).
I don't know what a const without a value means but I propose to let `const ONE: u8;` in the `system` file to just test the diagnostic message.

glihm commented on crates/dojo-lang/src/plugin_test_data/system:
```diff
@@ -84,6 +84,62 @@ mod withcomponent {
+trait IEmptyTrait;
+
+#[dojo::interface]
+trait IActions {
+    /// Even if it looks weird, the expected pattern for a constant in a trait is
+    /// `const <name>: <type>;`
+    /// See: https://github.com/starkware-libs/cairo/blob/600553d34a1af93ff7fb038ac677efdd3fd07010/crates/cairo-lang-parser/src/parser.rs#L835
+    const ONE: u8;
```
We may want to write an other test where everything works as expected, to ensure the normal behavior is well supported.

remybar replied:
Ok, I splitted this trait into 2 traits: one "faulty" with a constant inside, and one with correct functions only.

glihm reviewed (commented):
@remybar thanks for the update!
A small comment about testing.

Also, would you mind updating the `spawn-and-move` example with the new syntax? This example is usually taken as an example and showcasing the dojo stack. :)

remybar commented:
> Also, would you mind updating the `spawn-and-move` example with the new syntax? This example is usually taken as an example and showcasing the dojo stack. :)

Sure ! I updated the example 👍 

Thanks for the review !

glihm commented on crates/dojo-lang/src/plugin_test_data/system:
```diff
@@ -84,6 +84,67 @@ mod withcomponent {
+}
+
+#[dojo::interface]
+trait INominalTrait {
+    fn do_no_param();
+    fn do_no_param_but_self(self: @TContractState);
+    fn do_params(p1: dojo_examples::models::Direction, p2: u8);
+    fn do_params_and_self(self: @TContractState, p2: u8);
+    fn do_return_value(p1: u8) -> u16;
+    fn do_ref_self(ref self: TContractState);
+
+    #[my_attr]
+    fn do_with_attrs(p1: u8) -> u16;
+}
```
We still miss a non-faulty test case, no? The `ref self` and the `#[my_attr]` should go into the faulty one, or you're thinking of adding one more?

remybar replied:
Ah yes, I will move `ref self` to the faulty trait !

For `#[my_attr]`, from the code expansion point of view, this is not an error. If a function has some attributes, they will be copied. But from Dojo point of view, maybe it's strange to have some attributes on functions in a contract trait ? So, if you think it is better to move this case in the faulty trait, I will do it 👍

glihm commented on examples/spawn-and-move/src/actions.cairo:
```diff
@@ -83,8 +82,7 @@ mod actions {
 
-        fn move(self: @ContractState, direction: Direction) {
-            let world = self.world_dispatcher.read();
+        fn move(direction: Direction, world: IWorldDispatcher) {
```
Seeing this, we must inline and remove the `world: IWorldDispatcher` from arguments **if and only if it's the first argument**. We shouldn't allow two arguments with this type either (an other faulty one to add! hehe).

Because at ABI level, removing `world` at first position can be understood. If it's randomly placed in the list of inputs of the function, it can quickly become difficult to call the systems.

remybar replied:
Thanks ! I will update the code accordingly 👍

glihm commented on examples/spawn-and-move/src/actions.cairo:
```diff
@@ -1,15 +1,15 @@
+#[dojo::interface]
+trait IActions {
+    fn spawn();
+    fn move(direction: dojo_examples::models::Direction);
```
To be consistent, could we import the Direction and use it as `Position` and `Vec2` while we're editing the code here?
```suggestion
    fn move(direction: Direction);
```

remybar commented:
So, to summarize:

* **[dojo::interface]**
  * add `<TContractState>` to the trait name,
  * add `self` parameter if not already present as first parameter,
  * raise a diagnostic error if there is a `ref self` parameter in at least one function,
  * raise a diagnostic error if there is other things than functions in the trait.
 
* **[dojo::contract]**
  * add `self` parameter if not already present as first parameter,
  * remove the `world: IWorldDispatcher` parameter if it is the first parameter of the function (self excluded) and if there is no more than one parameter of type `IWorldDispatcher`,
  * inline the world parameter by adding the `let world = self.world_dispatcher.read();` as first function statement if this world parameter has been removed,
  * raise a diagnostic error if there is a `ref self` parameter in at least one function,
  * raise a diagnostic error if there is more than one parameter of type `IWorldDispatcher`.

glihm commented on crates/dojo-lang/src/contract.rs:
```diff
@@ -197,8 +200,232 @@ impl DojoContract {
+        let mut count = 0;
+
+        param_list.elements(db).iter().for_each(|param| {
+            let (_, _, param_type) = self.get_parameter_info(db, param.clone());
```
We should check that the `param_name` **is** `world` as it is expanded as such and we only authorize one `IWorldDispatcher`:
```rust
"let world = self.world_dispatcher.read();\n"
```

remybar replied:
In fact, the `check_world_dispatcher` function counts the number of parameters of type `IWorldDispatcher` and indicates if there is more than one.

Then, in `rewrite_parameters`, the `world: IWorldDispatcher` parameter is removed only if:
- there is only one `IWorldDispatcher` (which is `world`),
- `world` is the first parameter excluding `self` (so, if `self` is not there, index = 0, if `self` is already there, index=1).

Finally, if `world` has been removed, the statement to read `world` from `world_dispatcher` is added.

Maybe I should rename the `check_world_dispatcher` to something like `check_multiple_world_dispatchers`.

Is it fine for you now or did I miss something ? 🤔

glihm replied:
Yep, totally agree on the flow and it's great.

The only caveat is that if I use `ahah_world: IWorldDispatcher` instead of `world: IWorldDispatcher`, the compilation fails with the same cryptic error of wrong number of arguments. Because in the `rewrite_parameters` the `else` is not an error, just a re-write.

The naming is good to me. :+1:

remybar replied:
Ah yes, I see !
So the rule is: if there is a `IWorldDispatcher` parameter, it must be:
- the first parameter,
- named `world`,
- without any other `IWorldDispatcher` parameters.
👍

glihm replied:
> Ah yes, I see !
> 
> So the rule is: if there is a `IWorldDispatcher` parameter, it must be:
> 
> - the first parameter,
> 
> - named `world`,
> 
> - without any other `IWorldDispatcher` parameters.
> 
> 👍 

Nicely resumed, yes! Sorry for lack of precision on the first review round.

If one do want an additional world to interact with, they can always pass an address (in the interface too then), and build the dispatcher into the system. 👍

glihm commented on crates/dojo-lang/src/contract.rs:
```diff
@@ -197,8 +200,232 @@ impl DojoContract {
+    ) -> bool {
+        let mut count = 0;
+
+        param_list.elements(db).iter().for_each(|param| {
```
What about being more strict, which may improve user comprehension. If `self` is present, `world` must be at index 1. At index `0` otherwise.

I've testing this:
```rust
fn move(direction: Direction, world: IWorldDispatcher) ...
```
And the error is:
```
error: The number of parameters in the impl function `ActionsImpl::move` is incompatible with `IActions::move`. Expected: 2, actual: 3.
 --> /dojo/examples/spawn-and-move/src/actions.cairo[actions]:99:17
        fn move(self: @ContractState, direction: Direction, world: IWorldDispatcher) {
                ^*****************************************************************^

error: Wrong number of arguments. Expected 2, found: 3
 --> /dojo/examples/spawn-and-move/src/actions.cairo[actions][contract]:209:5
    ActionsImpl::move(@contract_state, __arg_direction, __arg_world);
    ^**************************************************************^

error: Compilation failed.
```
Which can be easier for the user with a diagnostic saying: `If the world dispatcher is used, it is expected to the first parameter.`.

We've to see about `self` if we need to mention it, because user should not have to use it anyway.

remybar replied:
Yes, I agree !

Because it could be confusing for the user:
- if the user adds `world` as first parameter, he **must not** add it in the dojo::interface as it will be removed,
- if the user adds `world` anywhere in the function but not as first parameter, he **must** add it in the dojo::interface to avoid the compilation error you mentioned.

In fact, this is an acceptable constraint because we have the same kind of constraint for `self` when writing a smart contract :)

remybar replied:
But what about backward compatibility ? :-/

glihm commented on crates/dojo-lang/src/plugin_test_data/system:
```diff
@@ -84,6 +84,134 @@ mod withcomponent {
+        'land'
+    }
+
+    #[external(v0)]
```
`abi(embed_v0)` is a bit more painful to use as it can't be used on standalone function. But `external(v0)` is expected to be deprecated. Would it be possible to switch to `abi(embed_v0)` in this PR?

remybar replied:
Hum, from what I understand, these attributes are not really important in this `system` file because they are not really used by the plugin to generate some specific code. But it is better to have at least one attribute on each functions to be sure that function attributes are well copied in the output code.

I've updated the code to use `abi(embed_v0)` on `Impl` block and `external(v0)` on free functions. What should we use instead of `external(v0)` for free functions ? Or do we have to put external/public functions in a `Impl` block to have them exported in the ABI file maybe ?

glihm commented on crates/dojo-lang/src/plugin_test_data/system:
```diff
@@ -84,6 +84,134 @@ mod withcomponent {
+            'land'
+        }
+
+        fn do_with_world_not_first(vec: Vec2, world: IWorldDispatcher) -> felt252 {
```
Isn't this faulty? :P
(and the one below)

Well.. it depends if the current impl is only considering the `world` as first position to be replaced, and the other ones to be kept.
But as mentioned earlier, and @tarrencev correct me if I'm wrong on that, we always want the `world: IWorldDispatcher` to be inlined, hence faulty IMHO.

remybar replied:
If we decide to force `world` to be the first parameter, yes, this is faulty 👍

glihm reviewed (commented):
Awesome work man, some comments here that I think should be almost the latest on my end. It looks great!

@tarrencev when commenting on the issue I didn't get well the fact that the `world` is only injected in the `impl`. The interface is super clean and the Devex is very nice.
